### PR TITLE
Faster JoinHash output assembly

### PR DESCRIPTION
**Pull request by mrzzzrm**
While trying to predict the costs for JoinHash, I noticed somewhat erratic performance (anything from 40ms to 1.5s for roughly the same number of input rows). This PR introduces a bit of caching to avoid redundant calculation of PosLists when joining Reference Tables.

TPCH performance nicely increases by 355% for TPCH-3 and 240% for TPCH-10

this branch:
```json
{
  "benchmarks": [
    {
      "cpu_time": 154903426,
      "items_per_second": 6.45563507080078,
      "iterations": 33,
      "name": "TPC-H 1",
      "real_time": 154903426,
      "time_unit": "ns"
    },
    {
      "cpu_time": 120605387,
      "items_per_second": 8.29150390625,
      "iterations": 42,
      "name": "TPC-H 3",
      "real_time": 120605387,
      "time_unit": "ns"
    },
    {
      "cpu_time": 345068374,
      "items_per_second": 2.89797639846802,
      "iterations": 15,
      "name": "TPC-H 5",
      "real_time": 345068374,
      "time_unit": "ns"
    },
    {
      "cpu_time": 2115483,
      "items_per_second": 472.705230712891,
      "iterations": 1000,
      "name": "TPC-H 6",
      "real_time": 2115483,
      "time_unit": "ns"
    },
    {
      "cpu_time": 816416877,
      "items_per_second": 1.22486436367035,
      "iterations": 7,
      "name": "TPC-H 9",
      "real_time": 816416877,
      "time_unit": "ns"
    },
    {
      "cpu_time": 193529734,
      "items_per_second": 5.16716480255127,
      "iterations": 26,
      "name": "TPC-H 10",
      "real_time": 193529734,
      "time_unit": "ns"
    }
  ],
  "context": {
    "benchmark_mode": "IndividualQueries",
    "build_type": "release",
    "chunk_size": 4294967295,
    "date": "2018-03-05 16:02:46",
    "scale_factor": 0.00999999977648258
  }
}
```

master:
```json
{
  "benchmarks": [
    {
      "cpu_time": 174671831,
      "items_per_second": 5.72502136230469,
      "iterations": 29,
      "name": "TPC-H 1",
      "real_time": 174671831,
      "time_unit": "ns"
    },
    {
      "cpu_time": 539360749,
      "items_per_second": 1.85404658317566,
      "iterations": 10,
      "name": "TPC-H 3",
      "real_time": 539360749,
      "time_unit": "ns"
    },
    {
      "cpu_time": 836498558,
      "items_per_second": 1.19545936584473,
      "iterations": 6,
      "name": "TPC-H 5",
      "real_time": 836498558,
      "time_unit": "ns"
    },
    {
      "cpu_time": 2085652,
      "items_per_second": 479.466247558594,
      "iterations": 1000,
      "name": "TPC-H 6",
      "real_time": 2085652,
      "time_unit": "ns"
    },
    {
      "cpu_time": 1888284852,
      "items_per_second": 0.529581129550934,
      "iterations": 3,
      "name": "TPC-H 9",
      "real_time": 1888284852,
      "time_unit": "ns"
    },
    {
      "cpu_time": 649493338,
      "items_per_second": 1.53966164588928,
      "iterations": 8,
      "name": "TPC-H 10",
      "real_time": 649493338,
      "time_unit": "ns"
    }
  ],
  "context": {
    "benchmark_mode": "IndividualQueries",
    "build_type": "release",
    "chunk_size": 4294967295,
    "date": "2018-03-05 16:08:28",
    "scale_factor": 0.00999999977648258
  }
}
```

### CODE REVIEW

**Comment by Bouncner** [comment]
I think you've pasted the same output twice?

**Comment by Bouncner** [comment]
And could you define for me what a "performance improvement of 240%" is? :)

**Comment by mrzzzrm** [comment]
"same output twice" https://www.youtube.com/watch?v=sP_l5aPww-4&feature=youtu.be&t=51s

If I can eat 1 apple per Minute and then improve to 5 apples then I improved by 400%. Is that how you do it in the science?

**Comment by Bouncner** [comment]
I need a real life proof. Of you, eating 5 apples/min.

File: src/lib/operators/join_hash.cpp [hunk 1]
```diff
@@ -698,55 +723,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
       }
       _output_table->emplace_chunk(std::move(output_chunk));
     }
 
     return _output_table;
   }
 
+  // See usage in _on_execute() for doc.
+  static PosListsByColumn setup_pos_lists_by_column(const std::shared_ptr<const Table> input_table) {
+    std::map<std::vector<const PosList*>, PosListsSPtr> shared_pos_lists_by_pos_lists;
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Why is it necessary / advantageous to use a raw pointer here?

File: src/lib/operators/join_hash.cpp [hunk 2]
```diff
@@ -698,55 +723,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+        write_output_chunks(output_chunk, _right_in_table, right_pos_lists_by_column, right, ref_col_right);
       }
       _output_table->emplace_chunk(std::move(output_chunk));
     }
 
     return _output_table;
   }
 
+  // See usage in _on_execute() for doc.
+  static PosListsByColumn setup_pos_lists_by_column(const std::shared_ptr<const Table> input_table) {
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: This gives you a list of all the PosLists for one column, correct? I don't yet understand why this helps.

File: src/lib/operators/join_hash.cpp [hunk 3]
```diff
@@ -698,55 +723,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
-        auto new_pos_list = std::make_shared<PosList>();
-        for (const auto& row : pos_list) {
-          if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
-            new_pos_list->push_back(row);
-          } else {
-            new_pos_list->push_back(input_pos_lists.at(row.chunk_id)->at(row.chunk_offset));
+        const auto& input_table_pos_lists = input_pos_list_ptrs_sptrs_by_column[column_id];
+
+        auto iter = output_pos_list_by_input_pos_list_ptrs_sptr.find(input_table_pos_lists);
+        if (iter == output_pos_list_by_input_pos_list_ptrs_sptr.end()) {
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Naming: I would call it `output_pos_list_cache` or something like that.

File: src/lib/operators/join_hash.cpp [hunk 4]
```diff
@@ -118,6 +118,10 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
   // Determine correct type for hashing
   using HashedType = typename JoinHashTraits<LeftType, RightType>::HashType;
 
+  using PosLists = std::vector<const PosList*>;
+  using PosListsSPtr = std::shared_ptr<PosLists>;
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: yap

File: src/lib/operators/join_hash.cpp [hunk 5]
```diff
@@ -685,6 +689,27 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
                              _right_in_table->get_chunk(ChunkID{0})->get_column(ColumnID{0}))
                              ? true
                              : false;
+    auto output_left_side = !_inputs_swapped || (_mode != JoinMode::Semi && _mode != JoinMode::Anti);
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: kk

File: src/lib/operators/join_hash.cpp (hunk 1 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: I did a lot of "lets get rid of ref counting" while optimizing this, which helped a little. I'll try to benchmark how much of a difference this makes.

File: src/lib/operators/join_hash.cpp (hunk 3 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: jap

File: src/lib/operators/join_hash.cpp (hunk 2 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: The key is that it returns a shared ptr to a PosList for each column, making it possible to identify columns in the input table that can share output pos lists.

File: src/lib/operators/join_hash.cpp [hunk 6]
```diff
@@ -685,6 +688,27 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+     *
+     * They do two things:
+     *      - Make it possible to re-use output pos lists if two columns in the input table have exactly the same
+     *          PosLists Chunk by Chunk
+     *      - Avoid creating the std::vector<const PosList*> for each Partition over and over again.
+     */
+    PosListsByColumn left_pos_lists_by_column;
+    PosListsByColumn right_pos_lists_by_column;
+
+    if (ref_col_left && only_output_right_input) {
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: This needs a comment on why this condition is relevant

File: src/lib/operators/join_hash.cpp [hunk 7]
```diff
@@ -698,55 +722,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
 
       // we need to swap back the inputs, so that the order of the output columns is not harmed
       if (_inputs_swapped) {
-        write_output_chunks(output_chunk, _right_in_table, right, ref_col_right);
+        write_output_chunks(output_chunk, _right_in_table, right_pos_lists_by_column, right, ref_col_right);
 
         // Semi/Anti joins are always swapped but do not need the outer relation
-        if (_mode != JoinMode::Semi && _mode != JoinMode::Anti) {
-          write_output_chunks(output_chunk, _left_in_table, left, ref_col_left);
+        if (only_output_right_input) {
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: I feel like we got the logic / naming wrong now.

File: src/lib/operators/join_hash.cpp [hunk 8]
```diff
@@ -698,55 +722,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+
+    for (ColumnID column_id{0}; column_id < input_table->column_count(); ++column_id) {
+      // Get all the input pos lists so that we only have to pointer cast the columns once
+      auto pos_list_ptrs = std::make_shared<PosLists>(input_table->chunk_count());
+      auto pos_lists_iter = pos_list_ptrs->begin();
+
+      for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); chunk_id++) {
+        // This works because we assume that the columns have to be either all ReferenceColumns or none.
+        const auto& ref_column =
+            *static_cast<const ReferenceColumn*>(input_chunks[chunk_id]->columns()[column_id].get());
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Why don't you do a `static_pointer_cast` and dereference that one?

File: src/lib/operators/join_hash.cpp [hunk 9]
```diff
@@ -685,6 +688,27 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+     * Two Caches to avoid redundant reference materialization for Reference input tables. As the there might be
+     *  quite a lot Partitions (>500 seen), input Chunks (>500 seen), and columns (>50 seen), this speeds up
+     *  write_output_chunks a lot()
+     *
+     * They do two things:
+     *      - Make it possible to re-use output pos lists if two columns in the input table have exactly the same
+     *          PosLists Chunk by Chunk
+     *      - Avoid creating the std::vector<const PosList*> for each Partition over and over again.
+     */
+    PosListsByColumn left_pos_lists_by_column;
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: `    // This holds one entry per column in the table, not per BaseColumn in a single chunk`

File: src/lib/operators/join_hash.cpp [hunk 10]
```diff
@@ -685,6 +688,27 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
                              _right_in_table->get_chunk(ChunkID{0})->get_column(ColumnID{0}))
                              ? true
                              : false;
+    auto only_output_right_input = !_inputs_swapped || (_mode != JoinMode::Semi && _mode != JoinMode::Anti);
+
+    /**
+     * Two Caches to avoid redundant reference materialization for Reference input tables. As the there might be
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: "the there"

File: src/lib/operators/join_hash.cpp [hunk 11]
```diff
@@ -698,55 +722,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+
+    const auto& input_chunks = input_table->chunks();
+
+    for (ColumnID column_id{0}; column_id < input_table->column_count(); ++column_id) {
+      // Get all the input pos lists so that we only have to pointer cast the columns once
+      auto pos_list_ptrs = std::make_shared<PosLists>(input_table->chunk_count());
+      auto pos_lists_iter = pos_list_ptrs->begin();
+
+      for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); chunk_id++) {
+        // This works because we assume that the columns have to be either all ReferenceColumns or none.
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: ... and we have checked that the first one is a reference column before we called this method.

File: src/lib/operators/join_hash.cpp [hunk 12]
```diff
@@ -698,55 +722,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
-          if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
-            new_pos_list->push_back(row);
-          } else {
-            new_pos_list->push_back(input_pos_lists.at(row.chunk_id)->at(row.chunk_offset));
+        const auto& input_table_pos_lists = input_pos_list_ptrs_sptrs_by_column[column_id];
+
+        auto iter = output_pos_list_cache.find(input_table_pos_lists);
+        if (iter == output_pos_list_cache.end()) {
+          // Get the row ids that are referenced
+          const auto& input_table_pos_lists_ref = *input_table_pos_lists;
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Why is this defined here when it is only used once somewhere down?

File: src/lib/operators/join_hash.cpp [hunk 13]
```diff
@@ -698,55 +722,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
+
+          auto new_pos_list = std::make_shared<PosList>(pos_list.size());
+          auto new_pos_list_iter = new_pos_list->begin();
+          for (const auto& row : pos_list) {
+            if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
+              *new_pos_list_iter = row;
+            } else {
+              *new_pos_list_iter = (*input_table_pos_lists_ref[row.chunk_id])[row.chunk_offset];
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: I think introducing `input_table_pos_lists_ref` actually makes it harder to understand what is happening here. Can you find a name for `(*input_table_pos_lists_ref[row.chunk_id])`?

File: src/lib/operators/join_hash.cpp (hunk 6 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: "left_pos_lists_by_column won't be used if left is data table or not being output at all"

File: src/lib/operators/join_hash.cpp (hunk 8 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: that increases the ref count, changed it to static_pointer_cast anyway.

File: src/lib/operators/join_hash.cpp (hunk 9 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: yap

File: src/lib/operators/join_hash.cpp (hunk 11 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: Added assert "DebugAssert(input_table->get_type() == TableType::References, "Function only works for reference tables");" and removed the comment

File: src/lib/operators/join_hash.cpp (hunk 12 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: I feared there would be cost involved with dereferencing the shared_ptr, but changed it

File: src/lib/operators/join_hash.cpp (hunk 13 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: Changed it to 
```
              const auto& referenced_pos_list = *(*input_table_pos_lists)[row.chunk_id];
              *new_pos_list_iter = referenced_pos_list[row.chunk_offset];
```

File: src/lib/operators/join_hash.cpp (hunk 10 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: fixed

File: src/lib/operators/join_hash.cpp (hunk 7 above)
Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: Fixed only_output_right_input and reversed usage to if(!only_output_right_input)

**Comment by mrzzzrm** [comment]
Addressed

File: src/lib/operators/join_hash.cpp [hunk 14]
```diff
@@ -687,24 +687,29 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
...
     PosListsByColumn left_pos_lists_by_column;
     PosListsByColumn right_pos_lists_by_column;
 
-    if (ref_col_left && only_output_right_input) {
+    // left_pos_lists_by_column won't be used if left is data table or not being output at all
+    if (ref_col_left && !only_output_right_input) {
       left_pos_lists_by_column = setup_pos_lists_by_column(_left_in_table);
     }
+
+    // right_pos_lists_by_column won't be used if right is data table
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: Ok, but then why do we set it up? Or in other words: Why do we have that check for the left side but not for the right side?

Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: the right side is always written to the output_table

Comment (mrks) [reply src/lib/operators/join_hash.cpp]: So it _will_ be used?

Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: right_pos_lists_by_column will NOT be used if the right input is a DATA table only if it is a REFERENCE table.

Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: wow, that sounds angry: It's in write_output_columns: the caches won't be used for data tables thus it makes no sense to fill them.

Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: I could refactor the JoinHash to have different write_output methods for data and reference inputs but then again, the entire JoinHash could be refactored....

Comment (mrks) [reply src/lib/operators/join_hash.cpp]: Ah, I think I didn't understand the term "data table". We don't use it anywhere in the code (yet).

If you change this to "right_pos_lists_by_column will only be used if right is a reference table", we're good to go.

Comment (mrzzzrm) [reply src/lib/operators/join_hash.cpp]: changed

**Pull request merged**